include/shmem_pack_helper.cuh

```diff
@@ -0,0 +1,228 @@
+                              arg.counter, NVSHMEM_SIGNAL_SET, arg.neighbor_ranks[3]);
+    return;
+  case 4:
+    nvshmem_putmem_signal_nbi(getShmemBuffer<1>(4, arg), getShmemBuffer<0>(4, arg), arg.bytes[4], arg.sync_arr + 5,
+                              arg.counter, NVSHMEM_SIGNAL_SET, arg.neighbor_ranks[4]);
+    return;
+  case 5:
+    nvshmem_putmem_signal_nbi(getShmemBuffer<1>(5, arg), getShmemBuffer<0>(5, arg), arg.bytes[5], arg.sync_arr + 4,
+                              arg.counter, NVSHMEM_SIGNAL_SET, arg.neighbor_ranks[5]);
+    return;
+  case 6:
+    nvshmem_putmem_signal_nbi(getShmemBuffer<1>(6, arg), getShmemBuffer<0>(6, arg), arg.bytes[6], arg.sync_arr + 7,
+                              arg.counter, NVSHMEM_SIGNAL_SET, arg.neighbor_ranks[6]);
+    return;
+  case 7:
+    nvshmem_putmem_signal_nbi(getShmemBuffer<1>(7, arg), getShmemBuffer<0>(7, arg), arg.bytes[7], arg.sync_arr + 6,
+                              arg.counter, NVSHMEM_SIGNAL_SET, arg.neighbor_ranks[7]);
+    return;
+  default: return;
+  }
+}
+
+// this currently operates on arg.sync_arr + 2 * QUDA_MAX_DIM, i.e. is used for exchange ghost
+template <typename Arg> __device__ inline void shmem_signalwait(int dim, int dir, bool wait, const Arg &arg)
+{
+
+  bool amLast = false;
+
+  __syncthreads(); // make sure all threads in this block arrived here
+  if (quda::target::thread_idx().x == 0 && quda::target::thread_idx().y == 0 && quda::target::thread_idx().z == 0) {
+    int ticket = arg.retcount_intra[0].fetch_add(1);
+    // currently CST order -- want to make sure all stores are done before and for the last block we need that
+    // all uses of that data are visible
+    amLast = (ticket == gridDim.x * gridDim.y * gridDim.z - 1);
+  }
+  auto my_tile = cg::coalesced_threads();
+
+  // This operation will be performed by only the
+  // first 32-thread tile of each block
+  if (my_tile.meta_group_rank() == 0) {
+    amLast = my_tile.shfl(amLast, 0);
+    my_tile.sync();
+  }
+  if (amLast && my_tile.meta_group_rank() == 0) {
+
+    for (int shmemidx = my_tile.thread_rank(); shmemidx < 8; shmemidx += my_tile.size()) {
+      dim = shmemidx / 2;
+      dir = shmemidx % 2;
+
+      if (!(getNeighborRank(2 * dim + dir, arg) < 0)) {
+        // send data over IB if necessary
+        if (getShmemBuffer<1, decltype(arg)>(shmemidx, arg) != nullptr) {
+          shmem_putbuffer_signal(shmemidx, arg);
+        } else {
+          nvshmemx_signal_op(arg.sync_arr + 2 * dim + (1 - dir), arg.counter, NVSHMEM_SIGNAL_SET,
+                             getNeighborRank(2 * dim + dir, arg));
+        }
+      }
+    }
+    if (my_tile.thread_rank() == 0) arg.retcount_intra[0].store(0); // this could probably be relaxed
+    my_tile.sync();
+    for (int shmemidx = my_tile.thread_rank(); shmemidx < 8; shmemidx += my_tile.size()) {
+      dim = shmemidx / 2;
+      dir = shmemidx % 2;
+      if (wait && !(getNeighborRank(2 * dim + dir, arg) < 0)) {
```

Comment:
Comment please: this is only called for IB-connected paths?

Reply:
No, the wait is called for all neighbors.